### DISCUSSION TITLE
Breaking changes in model interfaces

### DISCUSSION THREAD

**Discussion by mhajas** [keycloak/keycloak#25311] [Feedback]
During the development of features we often need to change model interfaces (like UserProvider, GroupProvider, etc.). The change itself is okay, however, we sometimes forget third parties may implement these interfaces and such changes break their implementations. I know the ideal solution would be to refactor interfaces to not require breaking changes so often but we are far from that at the moment. 

An example of such change may be adding parameters to a method or introducing a new method to some interface.

### Possible approaches

We should be more careful about watching these, however, we should first discuss what is the best approach to introduce breaking changes. 

1. In compile time -> We do the breaking change and let users find out the hard way either during compiling the extension with the new version (the better option but not sure how many people recompile extensions with each upgrade) or in runtime with an exception like `MethodNotFound` or something similar.
2. In runtime only -> We make sure the interface change is compilable with an older version (essentially we add new methods all the time and deprecate old ones if exist) but throw an exception if a new method is called in a default method
3. Force implementing a default method from the author-> It is usually possible to implement some default implementation providing the same functionality, however, often highly inefficient, for example, loading all users and filtering based on something not supported by existing methods (just an example).

### Proposal

Go with option `3` and include a message in `Release notes` saying that this method was introduced and it needs to be implemented by each extension otherwise, the default implementation may impact performance. We will include a `warn` level message saying this implementation is inefficient and needs to be overridden by this provider. We will also restrict how often the message is printed (e.g. once in 5 seconds) so we don't flood the logs with these warnings.

Then we will create an issue with milestone `current_major_keycloak_version + 2` where we would remove the default implementation and go with option 1. for the rest of the people who haven't implemented it yet. A release note message will be included too. 

### Disadvantage

It will take 6 months to get to the desired state of the interface.

**Comment by rh-iatanaso** [comment]
> include a message in `Release notes`

Any chance this could be an action item on itself. For example, as part of PR that affects model interfaces an update in the Release Notes is a must ?

**Reply by mhajas** [reply]
Do you mean like Github action which fails if there is a change in some of the interfaces and the release notes is not updated?

**Reply by rh-iatanaso** [reply]
I am not sure if that is possible, but that would be great

**Reply by mhajas** [reply]
If we have a list of interfaces that cannot be broken it would be possible. However, we cannot say which change is a breaking change and which is not. Like if we change a Javadoc we don't need to write anything to the release notes so we would need some mechanism to waive the failed Github actions which currently require some higher admin rights if I am not mistaken. 

What we can have now for User storage is the following implementation: https://github.com/keycloak/keycloak/blob/main/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/storage/BackwardsCompatibilityUserStorageTest.java

But this still requires a reviewer to make sure there are no changes in that implementation. Also, we have this only for User interfaces not for the others.

**Reply by pedroigor** [reply]
@mhajas I would say that for changes like you are describing we should go with #1. The reason is what you mentioned, why provide a default and inefficient (potentially buggy) if people should consider upgrading their extension to the proper/best solution to the problem?

I agree that otherwise, it will slow down involving our features. Perhaps, potentially making them a bit more complex as we have to keep two different behaviors.

**Reply by sventorben** [reply]
@pedroigor In contrast, not providing a default implementation may slow down adoption of latest Keycloak version. There are extensions that users rely on. Extension developers need time to adjust their extensions to latest changes. User need to wait for them so that extensions are compatible with latest KC version before they can update.

**Comment by stianst** [comment]
The golden rule we are trying to achieve is that for any breaking change there is at least two releases that doesn't affect backwards compatibility. Essentially giving people at two stable releases (roughly 6 months) where they can make the required changes.

Putting this into an example:

* Let's say KC 23 as MyAPI v1
* KC 24 is then introduced with MyAPI v2 (not backwards compatible with v1) and deprecates MyAPI v1
* Users of Keycloak can update their clients/providers/etc while running KC 24 and/or KC 25
* KC 26 can now remove MyAPI v1

We would want to come up with a standard process for all APIs here, not a specific approach for model interfaces only.

**Reply by danielFesenmeyer** [reply]
Why not using [semver](https://semver.org/)?
Then there would be no need to explain and consider an additional version concept.

Applied to your example:
- Let's say KC 23 as MyAPI v1
- KC 23.1.x is then introduced with MyAPI v2 (not backwards compatible with v1) and deprecates MyAPI v1
- Users of Keycloak can update their clients/providers/etc while running KC 23.1.x and/or KC 23.2.x
- KC 24 can now remove MyAPI v1

Furthermore, there are tools like these that help preventing API breaks:
- https://github.com/siom79/japicmp
- https://revapi.org/revapi-site/main/index.html

**Reply by mhajas** [reply]
@stianst It sounds the same as suggested in the proposal (I wrongly computed the needed to get to the desired state) so thanks for the confirmation.

@danielFesenmeyer This may be problematic. When we release a major version we create a branch `release/23.0` where we backport all regressions and security fixes. To follow your suggestion we would need to backport interface changes too. This would break our release cycle as we usually do only micro releases of the previous version.

**Reply by jonkoops** [reply]
Technically we already adhere to semantic versioning, as breaking changes are only ever introduced in major versions. I agree that we could start doing minor versions to introduce deprecations (not breaking changes yet).

**Reply by danielFesenmeyer** [reply]
@mhajas I don't know the details of the keycloak release process, including backporting etc. Probably that's why I don't completely understand the problem. What's the difference of having release branches (1) `release/23.0.x`, `release/23.1.x`, `release/23.2.x` and `release/23.x` (pointing to the same commit as latest minor version branch `release/23.2.x`) instead of (2) `release/23.x`, `release/24.x` and `release/25.x`? In either case, the regressions and security fixes will probably only be backported to the currently newest "compatible" branch:

- for (1): `release/23.x` and `release/23.2.x`, whereas those will point to the same commit
- for (2): `release/25.x`

**Reply by rh-iatanaso** [reply]
@jonkoops I am not sure, that is being strictly followed. Definition of what is breaking change could be useful because breaking change can mean different thing for different consumers.

**Comment by mposolda** [comment]
IMO it can be also important to consider the type of the interface/class where breaking changes are introduced. In general, I see 2 types of interfaces:

1. Interfaces, which people are supposed to implement when they create their own providers. This includes `UserStorageSPI` and related interfaces like `USerStorageProviderFactory`, `UserStorageProvider`, `UserQueryProvider`, `UserRegistrationProvider`, `UserLookupProvider` etc. Here we cannot simply add new methods (unless they are `default` methods).

2. Interfaces, which people are supposed to just "consume". So just invoke the methods of those interfaces in their own provider implementations. Most of the interfaces falls into this category. Those are classes like `KeycloakSession`, `RealmModel` and also `UserProvider` or `GroupProvider`, which you mentioned. IMO it is completely fine to introduce new methods on those interfaces. As long as people just consume those, introducing new method should not break anything for them. So I don't see any issue with adding new method without explicitly documenting it. However removing methods or updating signatures of existing methods is more problematic. And for those cases, we need to have "deprecation" period as mentioned above IMO.

**Reply by mhajas** [reply]
Should we then somehow distinguish between 1 and 2? We have `server-spi` and `server-spi-private` but for example `UserProvider` and `GroupProvider` are in `server-spi`.

**Reply by mposolda** [reply]
Maybe, but I am not 100% sure if we need to distinguish at that level? At this moment, `server-spi` contains all the classes, which people can consume in their own providers and also some classes, which they can implement. And we don't distinguish between them right now.

We have some distinctions done by `Spi.isInternal`, which specifies if the SPI is internal and hence people are not supposed to implement it. There are few Spi in `server-spi`, which return false and hence related factory and provider classes (like for instance `HostnameProviderFactory` or `HostnameProvider`) fall into category 1 and hence we cannot simply add non-default methods on them per my understanding. Same for `UserStorageProviderSpi`, which is in `keycloak-model-legacy`, but it is still supported SPI per my understanding.

On the other hand, if someone implements `UserProvider` (resp. `GroupProvider`), he is supposed to implement also `UserProviderFactory` (resp. `GroupProviderFactory`), but those factories are in `server-spi-private` and also corresponding SPIs `UserSpi` (resp. `GroupSpi`) are marked as internal. Hence looks that it is safe to assume that `UserProvider` and `GroupProvider` are not supposed to be implemented directly by people in their extensions?

Fact is that we don't document it though in the Server Developer Guide https://www.keycloak.org/docs/latest/server_development/index.html . So perhaps we might add some documentation to make it clearer what interfaces are supposed to implement and what just to consume? Is documentation sufficient to address this or do we need something more?

**Reply by pedroigor** [reply]
I always thought that internal and private SPIs were OK to break. We do need to write release notes with details about the breaking changes you introduce. 

I'm afraid that otherwise, we are going to slow down in delivering the feature as fully supported and with a stable API. See https://access.redhat.com/support/offerings/techpreview.

**Reply by mposolda** [reply]
Yes, the question here is, what is private and internal SPI? From what I know, anything in `keycloak-server-spi` is kind of public API AFAIK? and per my understanding, it is not easily to break it.

In this bullet point, I am trying to differentiate between SPI, which people are supposed to just consume (so adding new method is fine IMO) or implement (even adding new method might be an issue here).

**Reply by mposolda** [reply]
On the other hand, if we can have an exceptions for the cases where it makes sense and for example just do (1) (so just break and document it), it can be nice as sometimes deprecation period doesn't makes too much sense...

**Reply by sschu** [reply]
@pedroigor I thought the same. Aren't there even log warnings when starting Keycloak if you implement a private SPI?
As a side note, for https://github.com/keycloak/keycloak/pull/24012 we opted to do 1) (break compile) instead of 2) (throw runtime exception in default method) on the `RoleProvider`.

**Reply by sventorben** [reply]
From my point of view internal and private SPIs are already public.

Take for example the `AuthenticatorSPI` which is part of `keycloak-server-spi-private`. It is [marked as internal](https://github.com/keycloak/keycloak/blob/23.0.6/server-spi-private/src/main/java/org/keycloak/authentication/AuthenticatorSpi.java#L33).
Following your discussion you would not have any authenticator extension out there. I maintain two myself which have thousands of downloads per version.

The server developer guide even gives instructions on how to implement it: https://www.keycloak.org/docs/latest/server_development/#implementing-an-authenticator

Same holds true for `RealmResourceSPI`.

So, from my point of view these SPIs are already public, no matter if they are marked as internal or reside within a specific module. They already become (maybe unintentionally?) part of the larger Keycloak ecosystem.
